packages/ra-tree-ui-materialui/src/TreeNodeWithChildren.js

```diff
@@ -0,0 +1,129 @@
+        ]),
+    };
+
+    handleChange = () => {
+        const { toggleNode, node } = this.props;
+
+        toggleNode(node.id);
+    };
+
+    render() {
+        const {
+            basePath,
+            cancelDropOnChildren,
+            children,
+            classes,
+            isExpanded,
+            node,
+            resource,
+            toggleNode,
+            treeNodeComponent: TreeNode,
+            treeNodeWithChildrenComponent,
+            treeNodeContentComponent: TreeNodeContent,
+            ...props
+        } = this.props;
+
+        return (
+            <ExpansionPanel
+                classes={{
+                    root: classes.panel,
+                }}
+                elevation={0}
+                expanded={isExpanded}
+                onChange={this.handleChange}
+            >
+                <ExpansionPanelSummary
+                    classes={{
+                        content: classes.panelSummaryContent,
+                        expandIcon: classes.expandIcon,
+                        root: classes.panelSummary,
+                        expanded: classes.panelSummaryExpanded,
+                    }}
+                    expandIcon={<KeyboardArrowDown />}
+                >
+                    <TreeNodeContent
+                        key={`TreeNodeContent${node.id}`}
+                        basePath={basePath}
+                        node={node}
+                        resource={resource}
+                        cancelDropOnChildren={cancelDropOnChildren}
+                        {...props}
+                    >
+                        {children}
+                    </TreeNodeContent>
+                </ExpansionPanelSummary>
+                <ExpansionPanelDetails
+                    classes={{
+                        root: classes.panelDetails,
+                    }}
+                >
+                    <List dense>
+                        {node.children.map(child => (
+                            <TreeNode
+                                key={`TreeNode_${child.id}`}
+                                basePath={basePath}
+                                classes={classes}
+                                node={child}
+                                resource={resource}
+                                treeNodeComponent={TreeNode}
+                                treeNodeWithChildrenComponent={
+                                    treeNodeWithChildrenComponent
+                                }
+                                treeNodeContentComponent={TreeNodeContent}
+                                {...props}
+                            >
+                                {children}
+                            </TreeNode>
+                        ))}
+                    </List>
+                </ExpansionPanelDetails>
+            </ExpansionPanel>
+        );
+    }
+}
+
+const TreeNodeWithChildren = ({ isExpanded, ...props }) => (
+    <TreeContext.Consumer>
```

Comment:
I'm really not sure you need this context. Can't we pass the props down instead?

Reply:
Sure

Reply:
It was part of my experimentations regarding performances but it has fixed them